src/radiation/radiation_dust_system.hpp

```diff
@@ -0,0 +1,876 @@
+
+	auto dEg_dT = kappaPoverE * d_fourpiboverc_d_t;
+
+	result.J00 = 0.0;
+	result.J0g.fillin(1.0);
+	result.Jg0 = dEg_dT;
+	for (int g = 0; g < nGroups_; ++g) {
+		if (tau[g] <= 0.0) {
+			result.Jgg[g] = -std::numeric_limits<double>::infinity();
+		} else {
+			result.Jgg[g] = -1.0 * kappaPoverE[g] / tau[g] - 1.0;
+		}
+	}
+
+	return result;
+}
+
+// Compute the Jacobian of energy update equations for the gas-dust-radiation system. The result is a struct containing the following elements:
+// J00: (0, 0) component of the Jacobian matrix. = d F0 / d Egas
+// F0: (0) component of the residual. = Egas residual
+// Fg_abs_sum: sum of the absolute values of the each component of Fg that has tau(g) > 0
+// J0g: (0, g) components of the Jacobian matrix, g = 1, 2, ..., nGroups. = d F0 / d R_g
+// Jg0: (g, 0) components of the Jacobian matrix, g = 1, 2, ..., nGroups. = d Fg / d Egas
+// Jgg: (g, g) components of the Jacobian matrix, g = 1, 2, ..., nGroups. = d Fg / d R_g
+// Fg: (g) components of the residual, g = 1, 2, ..., nGroups. = Erad residual
+template <typename problem_t>
+AMREX_GPU_DEVICE auto RadSystem<problem_t>::ComputeJacobianForGasAndDustWithPE(
+    double T_gas, double T_d, double Egas_diff, quokka::valarray<double, nGroups_> const &Erad, quokka::valarray<double, nGroups_> const &Erad0,
+    double PE_heating_energy_derivative, quokka::valarray<double, nGroups_> const &Rvec, quokka::valarray<double, nGroups_> const &Src, double coeff_n,
+    quokka::valarray<double, nGroups_> const &tau, double c_v, double /*lambda_gd_time_dt*/, quokka::valarray<double, nGroups_> const &kappaPoverE,
+    quokka::valarray<double, nGroups_> const &d_fourpiboverc_d_t) -> JacobianResult<problem_t>
+{
+	JacobianResult<problem_t> result;
+
+	const double cscale = c_light_ / c_hat_;
+
+	result.F0 = Egas_diff;
+	result.Fg = Erad - Erad0 - (Rvec + Src);
+	result.Fg_abs_sum = 0.0;
+	for (int g = 0; g < nGroups_; ++g) {
+		if (tau[g] > 0.0) {
+			result.Fg_abs_sum += std::abs(result.Fg[g]);
+			result.F0 += cscale * Rvec[g];
+		}
+	}
+	result.F0 -= PE_heating_energy_derivative * Erad[nGroups_ - 1];
+
+	// const auto d_fourpiboverc_d_t = ComputeThermalRadiationTempDerivativeMultiGroup(T_d, radBoundaries_g_copy);
+	AMREX_ASSERT(!d_fourpiboverc_d_t.hasnan());
+
+	// compute Jacobian elements
+	// I assume (kappaPVec / kappaEVec) is constant here. This is usually a reasonable assumption. Note that this assumption
+	// only affects the convergence rate of the Newton-Raphson iteration and does not affect the converged solution at all.
+
+	const auto d_Eg_d_Rg = -1.0 * kappaPoverE / tau;
+
+	result.J00 = 1.0;
+	result.J0g.fillin(cscale);
+	if (tau[nGroups_ - 1] <= 0.0) {
+		result.J0g[nGroups_ - 1] = LARGE;
+	} else {
+		result.J0g[nGroups_ - 1] -= PE_heating_energy_derivative * d_Eg_d_Rg[nGroups_ - 1];
+	}
+	const double d_Td_d_T = 3. / 2. - T_d / (2. * T_gas);
+	// const double coeff_n = dt * dustGasCoeff_local * num_den * num_den / cscale;
```

Comment:
Same comment about the coeff_n comment as in the analogous line in the previous function.

Reply:
Same.